<<readBam, echo=FALSE, cache=TRUE>>=

## how many reads?
command <- "ls /home/ele/Data/RNAseq/ | grep --perl \"AA|AJ_\" | parallel \'echo {}; grep \"@\" /home/ele/Data/RNAseq/{}/*_1.solfastq.gz.fastq | wc -l\'"
nReads.raw <- readLines(pipe(command))
nReads <- as.data.frame(t(matrix(nReads.raw, nrow=2)))
names(nReads) <- c("library", "raw.reads")

## Rsamtools
what <- c("rname")# , "strand", "pos", "qwidth", "seq")
param <- ScanBamParam(what = what)

files <- list.files("/home/ele/Data/RNAseq/mapping/", "*.bam$")

## alternative method using library(GenomicRanges) would be:
## aligns <- readBamGappedAlignments("/home/ele/Data/RNAseq/mapping/AJ_T19M_1.bam")
## ie. bam <- scanBam("/home/ele/Data/RNAseq/mapping/AJ_T19M_1.bam", param=param)


countsList <- list()
for (fi in files) {
  bam <- scanBam(paste("/home/ele/Data/RNAseq/mapping/", fi, sep=""), param=param)
  counts <- table(bam[[1]]$rname)
  counts.frame <- as.data.frame(counts)
  countsList[[fi]] <- counts.frame
}

counts.frame.long <- melt(countsList, id.vars="Var1")
counts.frame.long$variable <- NULL

counts.frame.wide <- reshape(counts.frame.long, v.names="value",
                             idvar="Var1", direction="wide", timevar="L1")
names(counts.frame.wide) <- gsub("^value\\.", "", names(counts.frame.wide))
names(counts.frame.wide) <- gsub("\\.bam$", "", names(counts.frame.wide))
row.names(counts.frame.wide) <- counts.frame.wide$Var1
counts.frame.wide$Var1 <- NULL

## Just to see the clustering fo technical replicates!!!

sex.conds.raw <- factor(ifelse(grepl("M_\\d", names(counts.frame.wide)), "male", "female" ))

cds.raw <- newCountDataSet(counts.frame.wide, sex.conds.raw)
cds.raw <- estimateSizeFactors(cds.raw)
cds.raw <- estimateVarianceFunctions(cds.raw)

vsd.raw <- getVarianceStabilizedData( cds.raw )
dists.raw <- dist( t( vsd.raw ) )
idists.raw <- as.matrix(dists.raw)

### Sum up technical replicates!!!
counts.frame.long$L1 <- gsub("_\\d.bam", "", counts.frame.long$L1)
counts.frame.long$L1 <- as.factor(counts.frame.long$L1)
counts.frame <- as.data.frame(tapply(counts.frame.long$value,
                                     list(counts.frame.long$Var1, counts.frame.long$L1),
                                     sum))

mapped.raw <- colSums(counts.frame)

## use counts frame for all calculations
## remove outlier sample based on the plot below
## counts.frame$AJ_T26F <- NULL
### limit to good quality for the moment:
good.contigs <- contig.df[contig.df$AcMN, "contig"]
counts.frame.good <- counts.frame[ rownames(counts.frame)%in%good.contigs, ]

mapped.good <- colSums(counts.frame.good)

map.tab <- merge(nReads, mapped.raw, by.x="library", by.y="row.names")

map.tab <- merge(map.tab, mapped.good, by.x="library", by.y="row.names")

<<tech.rep.plot, echo=FALSE>>=
## check that all tech-reps are clustering, m/f are clustering, etc...
png("../figures/heat_tech_rep.png", width=720, height=720)
heatmap(idists.raw , symm=TRUE, margins = c (7,7))
dev.off()
@ 

